UI/window-dock-youtube-app.cpp

```diff
@@ -0,0 +1,386 @@
+				     streaming_mode_t mode)
+{
+	if (!dockBrowser)
+		return;
+	// update channelId if empty:
+	UpdateChannelId();
+	// notify Kiosk API:
+	std::string script;
+	if (mode == YTSM_ACCOUNT) {
+		script = QString(R"""(
+			if (window.location.hostname == 'studio.youtube.com') {
+				let event = {
+					type: '%1',
+					channelId: '%2',
+					videoId: '%3',
+				};
+				console.log(event);
+				if (window.ytlsapi && window.ytlsapi.dispatchEvent)
+					window.ytlsapi.dispatchEvent(event);
+			}
+		)""")
+				 .arg(event)
+				 .arg(channelId)
+				 .arg(video_id)
+				 .toStdString();
+	} else {
+		const char *stream_key = video_id;
+		script = QString(R"""(
+			if (window.location.hostname == 'studio.youtube.com') {
+				let event = {
+					type: '%1',
+					streamKey: '%2',
+				};
+				console.log(event);
+				if (window.ytlsapi && window.ytlsapi.dispatchEvent)
+					window.ytlsapi.dispatchEvent(event);
+			}
+		)""")
+				 .arg(event)
+				 .arg(stream_key)
+				 .toStdString();
+	}
+	dockBrowser->executeJavaScript(script);
+
+	// in case of user still not logged in in dock panel, remember last event
+	SetInitEvent(mode, event, video_id, channelId.toStdString().c_str());
+}
+
+void YouTubeAppDock::Update()
+{
+	std::string url = InitYTUserUrl();
+
+	if (!dockBrowser)
+		CreateBrowserWidget(url);
+	else {
+		dockBrowser->setURL(url);
+	}
+
+	// if streaming already run, let's notify YT about past event
+	if (OBSBasic::Get()->StreamingActive()) {
+		obs_service_t *service_obj = OBSBasic::Get()->GetService();
+		OBSDataAutoRelease settings =
+			obs_service_get_settings(service_obj);
+		if (IsUserSignedIntoYT()) {
+			channelId.clear(); // renew channelId
+			UpdateChannelId();
+			const char *broadcast_id =
+				obs_data_get_string(settings, "broadcast_id");
+			SetInitEvent(YTSM_ACCOUNT, INGESTION_STARTED,
+				     broadcast_id,
+				     channelId.toStdString().c_str());
+		} else {
+			const char *stream_key =
+				obs_data_get_string(settings, "key");
+			SetInitEvent(YTSM_STREAM_KEY, INGESTION_STARTED,
+				     stream_key);
+		}
+	} else
+		SetInitEvent(IsUserSignedIntoYT() ? YTSM_ACCOUNT
+						  : YTSM_STREAM_KEY);
+
+	dockBrowser->reloadPage();
+}
+
+void YouTubeAppDock::UpdateChannelId()
+{
+	if (channelId.isEmpty()) {
+		YoutubeApiWrappers *apiYouTube = GetYTApi();
+		if (apiYouTube) {
+			ChannelDescription channel_description;
+			if (apiYouTube->GetChannelDescription(
+				    channel_description)) {
+				channelId = channel_description.id;
+			} else {
+				blog(LOG_ERROR,
+				     "YT: AccountConnected() Failed to get channel id");
+			}
+		}
+	}
+}
+
+void YouTubeAppDock::SetInitEvent(streaming_mode_t mode, const char *event,
+				  const char *video_id, const char *channelId)
+{
+	const std::string version = App()->GetVersionString();
+	blog(LOG_DEBUG, "YT: obs version: %s", version.c_str());
+
+	QString api_event;
+	if (event) {
+		if (mode == YTSM_ACCOUNT) {
+			api_event = QString(R"""(,
```

Comment:
From what I could tell there will never be a situation in which `api_event` will be empty (it's either handling `YTSM_ACCOUNT` or anything else.

As such the string composition for `script` can assume that the `%4` placeholder will always be replaced by a string, so you can add the comma directly there:

```
console.log("initEvent:", {
 					initClientMode: client_mode,
 					%4 });
```

Adding the comma here is only necessary if `api_event` can be an empty string.

Reply:
> there will never be a situation in which `api_event` will be empty

No, it will be.
We have two use cases:
1. dock panel opened while streaming is active. in this case api_event is set to latest event (INGESTION_STARTED, for example) and YT backend notified about situation (with current broadcast id or stream key). api_event is not empty.  
   in fact we remembering last sent event and sending it with "init" event when dock panel reopened (reloaded).
2. dock panel opened initially, without any other user actions before. in this case no api_event is required and created and no additional information sent to YT backend. api_event is empty.

Lines [322](https://github.com/yuriy-chumak/obs-studio/blob/818c0aedc82dfe57413889d863a6784f88364e96/UI/window-dock-youtube-app.cpp#L322)-344 of window-dock-youtube-app.cpp

Reply:
This is now resolved.